File: Bridge/Bridge.java
class Shape {
    private Drawer drawer;
    
    public Shape(Drawer drawer) {
        this.drawer = drawer;
    }

    public <T extends Number> void draw(T... args) {
        this.drawer.draw(args);
    }
}

class CircleShape extends Shape {
    private int radius;
    
    public CircleShape(int radius) {
        super(new CircleDrawer());
        this.radius = radius;
    }

    public void draw() {
        super.draw(radius);
    }
}

class RectangleShape extends Shape {
    private int length;
    private int width;
    
    public RectangleShape(int length, int width) {
        super(new RectangleDrawer());
        this.length = length;
        this.width = width;
    }
    
    public void draw() {
        super.draw(length, width);
    }
}

interface Drawer {
    <T extends Number> void draw(T... args);
}

class CircleDrawer implements Drawer {
    public <T extends Number> void draw(T... args) {
        System.out.println("I am a circle, my radius " + 
            "is: " + args[0]);
    }
}

class RectangleDrawer implements Drawer {
    public <T extends Number> void draw(T... args) {
        System.out.println("I am a rectangle, my length" + 
            "is " + args[0] + ", width is " + args[1]);
    }
}

public class Bridge {
    public static void main(String[] args) {
        new CircleShape(100).draw();
        new RectangleShape(100, 200).draw();
    }
}

File: Bridge/Bridge.php
<?php
/**
 * Created by PhpStorm.
 * User: lanjialiang
 * Date: 2017/8/24
 * Time: 下午4:49
 */
interface DrawerInterface{
    public function draw();
}

class CircleDrawer implements DrawerInterface{
    public function draw(){
        return 'circle';
    }
}

class RectangleDrawer implements DrawerInterface{
    public function draw(){
        return 'rectangle';
    }
}
/***
 * bridge
 */
abstract class Shape implements DrawerInterface{
    /**
     * @var DrawInterface $drawer
     */
    protected $drawer = null;
    abstract public  function draw();
}

class CircleShape extends Shape{
    protected  $radius = 0;
    public function __construct($radius){
        $this->radius = $radius;
        $this->drawer = new CircleDrawer();
    }
    public function draw(){
        return $this->drawer->draw();
    }
}

class RectangleShape extends Shape{
    protected  $width=0;
    protected  $height =0;
    public function __construct($width,$height){
        $this->width = $width;
        $this->height = $height;
        $this->drawer = new RectangleDrawer();
    }

    public function draw(){
        return $this->drawer->draw();
    }
}

class Test{
    public function __invoke(){
        $circle = new CircleShape(1);
        assert($circle->draw() === 'circle','circleShape draw fail');
        $rectangle = new RectangleShape(10,10);
        assert($rectangle->draw() === 'rectangle','rectangleShape draw fail');
        print_r('draw success');
    }
}
$test = new Test();
$test();
